wallet.cc

```diff
@@ -164,13 +164,77 @@ std::string to_string(const PublicWebcash& epk)
     return webcash_string(epk.amount, "public", epk.pk);
 }
 
+// We group outputs based on their use.  There are currently four categories of
+// webcash recognized by the wallet:
+enum HashType : int {
+    // Pre-generated key that hasn't yet been used for any purpose.  To make
+    // backups possible and to minimize the chance of losing funds if/when
+    // wallet corruption occurs, the wallet maintains a pool of pre-generated
+    // secrets.  These are allocated and used, as needed, in FIFO order.
+    UNUSED = -1,
+
+    // Internal webcash generated either to redeem payments or mined webcash,
+    // change from a payment, or the consolidation of such outputs.  These
+    // outputs count towards the current balance of the wallet, but aren't shown
+    // explicitly.
+    CHANGE = 0,
+
+    // Outputs added via explicit import.  These are shown as visible, discrete
+    // inputs to the wallet.  The wallet always redeems received webcash upon
+    // import under the assumption that the imported secret value is still known
+    // to others or otherwise not secure.
+    RECEIVE = 1,
+
+    // Outputs generated via a mining report.  These are seen as visible inputs
+    // to a wallet, aggregated as "mining income."  The wallet always redeems
+    // mining inputs for change immediately after generation, in case the mining
+    // reports (which contain the secret) are made public.
+    MINING = 2,
+
+    // Outputs generated as payments to others.  These are intended to be
+    // immediately claimed by the other party, but we keep the key in this
+    // wallet in case there are problems completing the transaction.
+    PAYMENT = 3,
+};
+
+struct WalletSecret;
+
+struct WalletOutput {
+    int id;
+    uint256 hash;
+    std::unique_ptr<WalletSecret> secret;
+    int64_t amount;
+    bool spent;
+};
+
+struct WalletSecret {
+    int id;
+    uint256 secret;
+    bool mine;
+    bool sweep;
+};
+
 void Wallet::UpgradeDatabase()
 {
-    std::array<std::string, 1> tables = {
+    std::array<std::string, 3> tables = {
         "CREATE TABLE IF NOT EXISTS 'terms' ("
             "'id' INTEGER PRIMARY KEY NOT NULL,"
             "'body' TEXT UNIQUE NOT NULL,"
             "'timestamp' INTEGER NOT NULL);",
+        "CREATE TABLE IF NOT EXISTS 'output' ("
+            "'id' INTEGER PRIMARY KEY NOT NULL,"
+            "'timestamp' INTEGER NOT NULL,"
+            "'hash' BLOB UNIQUE NOT NULL,"
```

Comment:
This uniqueness constraint may overly constrain the wallet

Reply:
Right, because secrets can be reused. Thanks.